src/Orleans.Core/Core/ClusterClient.cs

```diff
@@ -61,6 +64,9 @@ public ClusterClient(IRuntimeClient runtimeClient, ILoggerFactory loggerFactory,
             {
                 participant?.Participate(clusterClientLifecycle);
             }
+
+            // It is fine for this field to be null in the case that the silo is not the host.
+            this.applicationLifetime = runtimeClient.ServiceProvider.GetService<IApplicationLifetime>() as ClientApplicationLifetime;
```

Comment:
This cast seems error prone.  If we need cast is ApplicationLifetime shouldn't we resolve that, and just register the same instance in container for ApplicationLifetime and IApplicationLifetime

Reply:
This is very intentional: we want the `IApplicationLifetime` implementation registered in the container but only if it's implemented by `ClientApplicationLifetime`. i.e, the lifetime is optional. We do not want to fire lifetime events if the host wasn't created by `ClientBuilder`. See above comment.